macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -2,125 +2,74 @@
 import numpy as np
+import logging
 
-import os, sys
-
-sys.path.append(os.getcwd())
-
-
-from macrosynergy.management.types import NoneType
+from macrosynergy.panel.historic_vol import expo_std, flat_std, expo_weights
+from macrosynergy.management.types import NoneType, QuantamentalDataFrame
 from macrosynergy.management.utils import (
     reduce_df,
     standardise_dataframe,
     is_valid_iso_date,
     qdf_to_ticker_df,
     ticker_df_to_qdf,
+    get_eops,
+    get_cid,
 )
-from macrosynergy.panel.historic_vol import get_cycles
 
+RETURN_SERIES_XCAT = "_PNL_USD1S_ASD"
+logger = logging.getLogger(__name__)
 
-def expo_weights(lback_periods: int = 21, half_life: int = 11):
-    """
-    Calculates exponential series weights for finite horizon, normalized to 1.
+cache = lru_cache(maxsize=None)
 
-    :param <int>  lback_periods: Number of lookback periods over which volatility is
-        calculated. Default is 21.
-    :param <int> half_life: Refers to the half-time for "xma" and full lookback period
-        for "ma". Default is 11.
 
-    :return <np.ndarray>: An Array of weights determined by the length of the lookback
-        period.
+@cache
+def flat_weights_arr(lback_periods: int, *args, **kwargs) -> np.ndarray:
+    """Flat weights for the lookback period."""
+    return np.ones(lback_periods) / lback_periods
 
-    Note: 50% of the weight allocation will be applied to the number of days delimited by
-          the half_life.
-    """
-    # Copied from macrosynergy.panel.historic_vol
-    decf = 2 ** (-1 / half_life)
-    weights = (1 - decf) * np.array(
-        [decf ** (lback_periods - ii - 1) for ii in range(lback_periods)]
-    )
-    weights = weights / sum(weights)
 
-    return weights
+@cache
+def expo_weights_arr(lback_periods: int, half_life: int, *args, **kwargs) -> np.ndarray:
+    """Exponential weights for the lookback period."""
+    return expo_weights(lback_periods=lback_periods, half_life=half_life)
 
 
-def expo_std(x: np.ndarray, w: np.ndarray, remove_zeros: bool = True) -> float:
-    """
-    Estimate standard deviation of returns based on exponentially weighted absolute
-    values.
-
-    :param <np.ndarray> x: array of returns
-    :param <np.ndarray> w: array of exponential weights (same length as x); will be
-        normalized to 1.
-    :param <bool> remove_zeros: removes zeroes as invalid entries and shortens the
-        effective window.
-
-    :return <float>: exponentially weighted mean absolute value (as proxy of return
-        standard deviation).
-
-    """
-    # Copied from macrosynergy.panel.historic_vol
-
-    assert len(x) == len(w), "weights and window must have same length"
-    if remove_zeros:
-        x = x[x != 0]
-        w = w[0 : len(x)] / sum(w[0 : len(x)])
-    w = w / sum(w)  # weights are normalized
-    mabs = np.sum(np.multiply(w, np.abs(x)))
-    return mabs
-
-
-def flat_std(x: np.ndarray, remove_zeros: bool = True) -> float:
+def _univariate_volatility(
```

Comment:
Are we using the `_univariate_volatility` function? If not, we should remove it.